### ISSUE TITLE
[4.x]: GraphQL cache key generation fails on second GraphQL call

### ISSUE THREAD

**Issue by Wiejeben** [craftcms/cms#11994]
### Description

After the rollout of a Craft 3 to Craft 4 update we started to notice some serialisation exceptions on generating the GraphQL cache key. Because of this GraphQL will only be cached for the first query, but not for the second one. After refreshing the second cache key does get generated because the first query was indeed cached.

The exception does not show up to our users, it just gets logged and `null` gets returned as the cache key.

Stacktrace:
```
[22-Sep-2022 11:50:00 Europe/Amsterdam] An Error occurred while handling another error:
Exception: Serialization of 'Closure' is not allowed in /var/www/vendor/craftcms/cms/src/services/Gql.php:1269
Stack trace:
#0 /var/www/vendor/craftcms/cms/src/services/Gql.php(1269): serialize(Array)
#1 /var/www/vendor/craftcms/cms/src/services/Gql.php(495): craft\services\Gql->_getCacheKey(Object(craft\models\GqlSchema), '{ping}', NULL, Array, NULL, NULL)
#2 /var/www/vendor/craftcms/cms/src/web/twig/Extension.php(1350): craft\services\Gql->executeQuery(Object(craft\models\GqlSchema), '{ping}', NULL, NULL)
#3 /var/www/storage/runtime/compiled_templates/39/394088f95faa5fe3182c0fbd02cff7dd.php(138): craft\web\twig\Extension->gqlFunction('{ping}')
#4 /var/www/vendor/twig/twig/src/Template.php(171): __TwigTemplate_d8e662eda5db898ad62e44b41f7ae499->block_head(Array, Array)
#5 /var/www/storage/runtime/compiled_templates/39/394088f95faa5fe3182c0fbd02cff7dd.php(46): Twig\Template->displayBlock('head', Array, Array)
#6 /var/www/vendor/twig/twig/src/Template.php(394): __TwigTemplate_d8e662eda5db898ad62e44b41f7ae499->doDisplay(Array, Array)
#7 /var/www/vendor/twig/twig/src/Template.php(367): Twig\Template->displayWithErrorHandling(Array, Array)
#8 /var/www/storage/runtime/compiled_templates/f7/f719560bf356996db7a0590c2f205865.php(49): Twig\Template->display(Array, Array)
#9 /var/www/vendor/twig/twig/src/Template.php(394): __TwigTemplate_b45bb191ab5bd0776a6c0dbc34631c67->doDisplay(Array, Array)
#10 /var/www/vendor/twig/twig/src/Template.php(367): Twig\Template->displayWithErrorHandling(Array, Array)
#11 /var/www/vendor/twig/twig/src/Template.php(379): Twig\Template->display(Array)
#12 /var/www/vendor/twig/twig/src/TemplateWrapper.php(40): Twig\Template->render(Array, Array)
#13 /var/www/vendor/twig/twig/src/Environment.php(277): Twig\TemplateWrapper->render(Array)
#14 /var/www/vendor/craftcms/cms/src/web/View.php(428): Twig\Environment->render('404', Array)
#15 /var/www/vendor/craftcms/cms/src/web/View.php(481): craft\web\View->renderTemplate('404', Array)
#16 /var/www/vendor/craftcms/cms/src/web/TemplateResponseFormatter.php(56): craft\web\View->renderPageTemplate('404', Array, 'site')
#17 /var/www/vendor/yiisoft/yii2/web/Response.php(1095): craft\web\TemplateResponseFormatter->format(Object(craft\web\Response))
#18 /var/www/vendor/craftcms/cms/src/web/Response.php(286): yii\web\Response->prepare()
#19 /var/www/vendor/yiisoft/yii2/web/Response.php(339): craft\web\Response->prepare()
#20 /var/www/vendor/yiisoft/yii2/web/ErrorHandler.php(136): yii\web\Response->send()
#21 /var/www/vendor/craftcms/cms/src/web/ErrorHandler.php(192): yii\web\ErrorHandler->renderException(Object(yii\web\NotFoundHttpException))
#22 /var/www/vendor/yiisoft/yii2/base/ErrorHandler.php(135): craft\web\ErrorHandler->renderException(Object(yii\web\NotFoundHttpException))
#23 /var/www/vendor/craftcms/cms/src/web/ErrorHandler.php(66): yii\base\ErrorHandler->handleException(Object(yii\web\NotFoundHttpException))
#24 [internal function]: craft\web\ErrorHandler->handleException(Object(yii\web\NotFoundHttpException))
#25 {main}
```

### Steps to reproduce

1. Execute two GraphQL query, for example in Twig:
```twig
{% set data = gql('{ping}') %}
{% set data2 = gql('{ping}') %}
```
2. Check the `phperrors.log`

### Expected behavior

Both queries to be cached.

### Actual behavior

Only first query gets cached.

### Craft CMS version

4.2.4

### PHP version

8.1.7

### Operating system and version

Docker alpine PHP 8.1

### Database type and version

MySQL 8.0.28

### Image driver and version

Imagick 3.7.0 (ImageMagick 7.1.0-16)

### Installed plugins and versions

-

**Comment by brianjhanson** [comment]
I'm having trouble replicating this one, but I noticed the line throwing the error is related to a `Craft::$app->getSites()->getCurrentSite()->id` call, is it possible there's something weird going on there?

I also see this is a request to render the 404 page, is this only happening on that template? 

Since this error was thrown after another error, what was the first error? It should be output right before this block in the logs.

**Comment by Wiejeben** [comment]
Just found out a bit more about the cause. The 404 inside the stacktrace can be ignored, that was just me testing.

The direct cause of the current exception I can lead back to https://github.com/sebastian-lenz/craft-linkfield/ however it does look like an issue caused by Craft.

I did a `var_dump(...)` of the `$context` variable on the first GraphQL execution and second and I noticed that `["_blockTypes":"craft\fields\Matrix":private]=>` changes from `null` to an array with `["_blockTypes":"craft\fields\Matrix":private]=>` (Can share on Discord since it might contain sensitive information).

The exception itself is caused by a `Closure` in `\lenz\linkfield\fields\LinkField::rules`, however changing the method to `return parent::rules();` still causes the exception.

Diving deeper, it looks like its caused by`\yii\validators\RequiredValidator`'s `when` property specifically:
```
["when"]=>
object(Closure)#460 (1) {
  ["this"]=>
  *RECURSION*
}
```

**Comment by brianjhanson** [comment]
After a bit of digging, I've tracked down the `$context` changes, and I think they're necessary and probably not related to the original serialization error.

The first time a request comes through, we miss the [`craft\services\Gql::getCacheResult()`](https://github.com/craftcms/cms/blob/deab3ac74f88d5bb35142d5aee1eaa9722b313fd/src/services/Gql.php#L497) and go on to generate the cache. Through the process of generating that cache, the matrix blocks types end up being loaded so they can be put in the schema. This results in the next query getting a different context (one with block types loaded in), which results in a different `cacheKey` and a miss to the cache. However, this difference is only in memory so on future queries, we will hit the cache before the blocks are loaded in and hit the cache on both queries. 

TL;DR the `$context` is changed through the process of generating the cache, but that should only happen on the first request after the graphql caches are cleared. 

So that doesn't explain the serialization error you're seeing. I'm still curious what error was thrown directly before that since the message is "An Error occurred while handling another error". Is there anything else in the logs that might be useful to track that down?

**Comment by brandonkelly** [comment]
Fixed for Craft 4.3 via #12086

**Comment by brandonkelly** [comment]
Craft 4.3 was released earlier this week with that fix.
